static/js/containers/ChannelPage.js

```diff
@@ -101,20 +61,37 @@ class ChannelPage extends React.Component {
       )
     }
   }
+}
 
-  render() {
-    const { channels, postsForChannel } = this.props
-    return <Loading restStates={[channels, postsForChannel]} renderContents={this.renderContents} />
-  }
+const shouldUpdateRequirements = R.complement(R.eqProps("channelName"))
+
+const updateRequirements = ({ dispatch, subscribedChannels, channelName }) => {
+  Promise.all([
+    dispatch(actions.channels.get(channelName)),
+    dispatch(actions.postsForChannel.get(channelName)).then(({ posts }) => {
+      dispatch(setPostData(posts))
+    }),
+    subscribedChannels ||
+      dispatch(actions.subscribedChannels.get()).then(channels => {
+        dispatch(setChannelData(channels))
+      })
+  ])
 }
 
-const mapStateToProps = state => {
+const mapStateToProps = (state, ownProps) => {
+  const channelName = getChannelName(ownProps)
+  const postIds = state.postsForChannel.data.get(channelName) || []
+  const channel = state.channels.data.get(channelName)
   return {
-    channels:           state.channels,
-    postsForChannel:    state.postsForChannel,
-    posts:              state.posts,
-    subscribedChannels: state.subscribedChannels
+    channelName,
+    channel,
+    posts:              safeBulkGet(postIds, state.posts.data),
+    subscribedChannels: state.subscribedChannels.loaded
+      ? safeBulkGet(state.subscribedChannels.data, state.channels.data)
+      : null,
+    loaded:  R.none(R.isNil, [channel]),
+    errored: anyError([state.channels, state.posts, state.subscribedChannels])
   }
 }
 
-export default connect(mapStateToProps)(ChannelPage)
+export default connect(mapStateToProps)(withLoading(ChannelPage, shouldUpdateRequirements, updateRequirements))
```

Comment:
That's much cleaner, although the args to `R.compose` needed to be the reverse of what you suggested since `withLoading` needs the props from `connect`.